infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java

```diff
@@ -251,11 +249,6 @@ private static void startServer(String[] args) {
                 System.exit(EXIT_ERROR_CODE);
             }
 
-            if (cleanupServer) {
-                clearDataFiles(serverContext);
-                cleanupServer = false;
-            }
-
```

Comment:
Move this logic to be executed by a handler thread and not the main thread can introduce a race. Although not clean, this race doesn't seem to be present on master.

The handler thread signaled to the main thread to shutdown and do the clean up.

But what can happen here is, as you are deleting the state, the main thread will break out of the `activeServer.startAndListen()`  and try to create a new server context while the data is being deleted by the handler thread in the `restartServer` method.

Reply:
The previous version also had a similar problem - we delete data and reset streamlog in parallel at the same time.

Fixed, by introducing and using `CorfuServerStateMachine` which brings total ordering between start, stop, reset_log_unit, shutdown operations